smart-contracts/contracts/Watcher.sol

```diff
@@ -0,0 +1,313 @@
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "./interfaces/ISovrynSwapNetwork.sol";
+import "./interfaces/ISovrynProtocol.sol";
+import "./interfaces/IPriceFeeds.sol";
+import "./interfaces/IWRBTCToken.sol";
+
+
+contract Watcher is AccessControl {
+    address public immutable RBTC_ADDRESS = address(0);
+    bytes32 public constant ROLE_EXECUTOR = keccak256("EXECUTOR");
+    bytes32 public constant ROLE_OWNER = DEFAULT_ADMIN_ROLE;
+
+    ISovrynProtocol public sovrynProtocol;
+    ISovrynSwapNetwork public sovrynSwapNetwork;
+    IPriceFeeds public priceFeeds;
+    IWRBTCToken public wrbtcToken;
+
+    event Arbitrage(
+        address indexed _sourceToken,
+        address indexed _targetToken,
+        uint256 _sourceTokenAmount,
+        uint256 _targetTokenAmount,
+        uint256 _priceFeedAmount,
+        uint256 _profit,
+        address _sender
+    );
+
+    event Liquidation(
+        bytes32 _loanId,
+        address indexed _loanToken,
+        address indexed _seizedToken,
+        uint256 _closeAmount,
+        uint256 _seizedAmount,
+        address _sender
+    );
+
+    event Swapback(
+        bytes32 _loanId,
+        address indexed _sourceToken,
+        address indexed _targetToken,
+        uint256 _sourceTokenAmount,
+        uint256 _targetTokenAmount,
+        address _sender
+    );
+
+    constructor(
+        ISovrynProtocol _sovrynProtocol,
+        ISovrynSwapNetwork _sovrynSwapNetwork,
+        IPriceFeeds _priceFeeds,
+        IWRBTCToken _wrbtcToken
+    ) {
+        sovrynProtocol = _sovrynProtocol;
+        sovrynSwapNetwork = _sovrynSwapNetwork;
+        priceFeeds = _priceFeeds;
+        wrbtcToken = _wrbtcToken;
+
+        // ROLE_OWNER is also the default admin role
+        _setupRole(ROLE_OWNER, msg.sender);
+    }
+
+    receive() external payable {
+        require(
+            msg.sender == address(wrbtcToken) ||
+            msg.sender == address(sovrynProtocol) ||
+            msg.sender == address(sovrynSwapNetwork),
+            "Watcher: only known contracts can transfer RBTC"
+        );
+    }
+
+    function arbitrage(
+        IERC20[] calldata _conversionPath,
+        uint256 _amount,
+        uint256 _minProfit
+    )
+    external
+    onlyRole(ROLE_EXECUTOR)
+    {
+        (IERC20 sourceToken, IERC20 targetToken, uint256 targetTokenAmount) = swapInternal(
+            _conversionPath,
+            _amount,
+            1
+        );
+
+        uint256 priceFeedReturn = priceFeeds.queryReturn(address(sourceToken), address(targetToken), _amount);
+        uint256 profit = targetTokenAmount - priceFeedReturn;
+        require(profit >= _minProfit, "Watcher: minimum profit not met");
```

Comment:
Replacing this check with `_minReturn` change is a good point -- will do the changes as requested.

"Arbitrage" in the usage of this contract is maybe not 100% accurate, since just means buying or selling a token at a price that's better than the price from an external source (price feed). In actual arbitrage, you would need to do the opposite action (sell/buy) on another exchange, or complete a loop and end up with the same asset (as suggested). Perhaps the require check is unnecessary.

However this implementation is consistent with the way arbitrage works in sovryn-node v1.